### PR TITLE
Create and use DataFlash_Class::should_log

### PR DESCRIPTION
**Pull request by peterbarker**
This also includes the subject of another commit which removes DataFlash_SITL, and some fixes for CardInserted which are part of one of @khancyr's cleanup patches.

### CODE REVIEW

File: libraries/DataFlash/DataFlash.h [hunk 1]
```diff
@@ -288,4 +289,7 @@ class DataFlash_Class
 
     // possibly expensive calls to start log system:
     void Prep();
+
+    bool _writes_enabled = false;
```
Comment (OXINARF) [inline libraries/DataFlash/DataFlash.h]: *Code size reduction squad*: false not needed.

File: APMrover2/Log.cpp [hunk 2]
```diff
@@ -510,9 +510,9 @@ void Rover::Log_Write_Vehicle_Startup_Messages()
 // start a new log
 void Rover::start_logging()
 {
-    in_mavlink_delay = true;
+    DataFlash.EnableWrites(false);
     DataFlash.StartUnstartedLogging();
-    in_mavlink_delay = false;
+    DataFlash.EnableWrites(true);
```
Comment (OXINARF) [inline APMrover2/Log.cpp]: I know this is almost a NFC, but I wonder why it is needed here - I don't see it in any other vehicle.

File: AntennaTracker/system.cpp [hunk 3]
```diff
@@ -253,10 +253,10 @@ void Tracker::check_usb_mux(void)
  */
 bool Tracker::should_log(uint32_t mask)
 {
-    if (in_mavlink_delay) {
+    if (!(mask & g.log_bitmask)) {
         return false;
     }
-    if (!(mask & g.log_bitmask)) {
+    if (!DataFlash.should_log()) {
```
Comment (OXINARF) [inline AntennaTracker/system.cpp]: I think this doesn't work as Tracker doesn't set armed state.

File: APMrover2/Log.cpp (hunk 2 above)
Comment (peterbarker) [inline APMrover2/Log.cpp]: Yeah, there are weirder inconsistencies I'm glossing over ATM, too ;-)

Comment (peterbarker) [inline APMrover2/Log.cpp]: ... also, while all of this is a little clean-uppy... Consider that currently library functions indirectly call WritePrioritisedBlock - meaning they're no doing the whole "am I doing the mavlink delay thing".  I'm not exactly sure how terribly bad that is, but with these changes they can start doing `DataFlash.should_log()`...

File: AntennaTracker/system.cpp (hunk 3 above)
Comment (peterbarker) [inline AntennaTracker/system.cpp]: It does now :-)

**Comment by peterbarker** [comment]
@OXINARF Thanks for the review.  And I have a whole heap of stuff coming.  Not all of which I'm currently convinced of.  Essentially I'm trying to get in_log_download to be a method on DataFlash.

**Comment by OXINARF** [comment]
Merged, thanks!

> And I have a whole heap of stuff coming. Not all of which I'm currently convinced of. Essentially I'm trying to get in_log_download to be a method on DataFlash.

That sounds good :grinning:

**Pull request merged**